### ISSUE TITLE
Add kmsKeyRef value for Bucket CRD

### ISSUE THREAD

**Issue by philchristensen** [aws-controllers-k8s/community#1649]
**Is your feature request related to a problem?**
I'm trying to create a Helm chart to deploy an S3 bucket that follows our company security credentials. For each bucket, we create a dedicated KMS key, but I'm not currently able to do that in a single Helm installation.

The Bucket CRD only has a `encryption/rules/applyServerSideEncryptionByDefault/kmsMasterKeyID`  field, and I need to be able to pass a `kmsKeyRef` object, much like I can when creating a DBInstance object for RDS.

**Describe the solution you'd like**
I'd like the Bucket CRD to support `encryption/rules/applyServerSideEncryptionByDefault/kmsKeyRef` so I can pass in the key object I'm creating.

**Describe alternatives you've considered**
All the examples I can find online depend on using Bash variables to combine k8s manifests with AWS CLI lookups, but I don't have access to these things in an automated, GitOps-based installation process.

**Comment by jaypipes** [comment]
This should be a relatively straightforward thing to add, right @RedbackThomson?

**Comment by jessebye** [comment]
Can this be reopened? We have the same use case.

**Comment by jessebye** [comment]
@jaypipes @a-hilaly we may be able to contribute towards a fix here, but I need a bit of guidance on where to start. This is impacting a use case we have where we create a new KMS key for each bucket and want to reference that KMS key's ARN in the bucket policy (as well as an IAM policy).

**Comment by michaelhtm** [comment]
Hello!
Thank you for taking initiative to contribute. To allow resource references please follow the following steps: https://aws-controllers-k8s.github.io/community/docs/contributor-docs/code-generator-config/#references-making-a-field-refer-to-another-resource

Feel free to reach out with any questions :)

**Comment by starlightromero** [comment]
> I took a attempt at giving this an implementation but there are some complications that make this feature non trivial. Since the reference is under the array encryption rules, we run into a bunch of edge cases of needing to match the response from the AWS SDK to the defined references. If the SDK response is modified it could be tricky to find how to update the object in a sane way.
> 
> > Meanwhile, Is there a workaround on how to dynamically pass key ARN to kmsMasterKeyID field within the Bucket object?
> 
> One workaround would be to use kro and define an RGD that looks at the status condition of the KMS key.

Good point. Unfortunately, at the current stage kro doesn't provide all the features my org needs. I'm tracking some open issues and looking forward to trying kro out once those issues are resolved
